UWP: Store Downloads ISO/CSO

Antx64 opened the pull request:
This PR contains more flexibility to the store downloads to allow some other content to be extracted from zip like iso and cso while also retaining EBOOT.pbp. I removed the requirement for it to have directory structure, but if that's still desired or a requirement, let me know and I'll fix it right away.

Also, this contains language support, fonts and had to remove gl shaders(.vsh files) so it could compile for UWP. I don't know if shaders will work, but at least they are included. Fixed a problem that prevented ppsspp gold from building uwp app.

Antx64 committed:
parent 2da9de9fe70d386bbe77c85e4dd6705a64c5167b
author Antx64 <46533123+Antx64@users.noreply.github.com> 1557556717 -0400
committer Antx64 <46533123+Antx64@users.noreply.github.com> 1557689648 -0400

Reverted files to original master hash

This commit includes the ability to download iso/cso, deleted useless binary files from UWP/Content folder, since they are now beign correctly copied from the assets folder and should no longer just update files. Fixed gold build, fixes submodules alter, and also, fixes the format change that occured thanks to copy paste and VS. Deleted .empty file from content folder accidentaly and had to recommit. Hurray for megit status

LunaMoo commented:
Blah what a mess.:|
 
 Duplicating common and large binary files on it's own - big NO, if the UWP build can't compile without assets in specific place, you should instead create some script to copy those files for you. Also custom shaders will not compile without their vsh files.

 The whole thing is outdated and tries to push outdated submodules over.


 The only real change in the code here is about ISO/CSO for homebrew store and that's unnecessary since homebrew never used any of it.
 Those formats are just containers, not all containers PPSSPP supports, but all which are used for warez stuff so I guess the point is to allow commercial games aka to allow piracy via some custom store server as discussed in some UWP specific issue where some people(you? I don't remember) begged for it talking exactly about custom homebrew store with pirated games.
 This ISN'T following this project spirit and I can't see a valid reason to really care about it. To begin with we don't need homebrew store to support anything more than Henrik wants to keep there as the point of it was to provide some popular(but legal) homebrew to be easily available which also brings a piece of PSP scene to people that never had a PSP nor PSP games. If someone wants a custom homebrew store for other reasons, they should just use their own modified build.

dominater01 commented:
@LunaMoo i have got word from hrydgard that for the store he was willing to make this change but instead antx64 has done it already

![Untitled](https://user-images.githubusercontent.com/8470028/57578525-36feeb00-745c-11e9-8979-01d894fb9f30.png)

hrydgard commented:
Adding support for automatically unzipping ISOs doesn't necessarily mean piracy so that part I'm OK with. I haven't looked at the code yet though, will do so later.

As for the UWP files, I fully agree with LunaMoo that we need a better solution than just duplicating the files, ideally a pre-build action to copy the files over.

Antx64 commented:
I added a commit that removed the copied binary files, and instead, included a symlink to each directory that contains the respective binaries. Can we compromise on that? As for the unzipping code, I just thought out a better implementation to avoid conflicts regarding the contents of the ppsspp store. As it is, it may unpack a lot of junk of the zip file when eboots are included. But please, do let me know if it's good enough as commited and updated later, or if you prefer to be updated on this same PR. Thanks

unknownbrackets reviewed (changes requested):
I'm not a big fan of using symlinks in git.

Why not just add a custom build step to UWP only that copies the assets from their primary location to where ever they need to be for UWP (and appropriate gitignore rules)?  This would mirror what other platforms do, like iOS, Linux, etc.

Also, the changes to the submodules here all look unintentional.  Also, removing a file in a future commit doesn't "remove" it since git remembers forever.  It will bloat the repository size.  To fix this, please rebase following this guide:
https://github.com/servo/servo/wiki/Beginner%27s-guide-to-rebasing-and-squashing

You'll want to make it so no commit ever:
1. Adds the large files you've since removed.
2. Changes the submodules/etc. that you unintentionally changed (since you're rebasing anyway.)

History is great and useful, but remembering mistakes just adds fluff and confusion (it can cause "blamespam" by making the history show you as the most recent author of something you changed by accident, and then reverted.)

-[Unknown]

Antx64 commented on UWP/glslang_UWP/glslang_UWP.vcxproj:
```diff
@@ -479,7 +479,7 @@
     <ClCompile Include="..\..\ext\glslang\SPIRV\SpvBuilder.cpp" />
     <ClCompile Include="..\..\ext\glslang\SPIRV\SPVRemapper.cpp" />
-    <ClCompile Include="..\..\ext\glslang\SPIRV\SpvPostProcess.cpp" />
+    <!--<ClCompile Include="..\..\ext\glslang\SPIRV\SpvPostProcess.cpp" />-->
```
I had to comment that, because at least when I updated the code, the build was complaining about not being able to find that file

Antx64 replied:
Got it, I'll keep the rebase in mind. Thanks, and sorry about that. I would've done a script to copy files, but didn't for two reasons:

1. VS/UWP needs this to be added as content.
2. I planned to allocate more time into this project next week to work on some other features I'm planning, by then I will do other PR to fix/improve some of this things and make PR for the new features separately.

Antx64 replied:
WTF? I didn't mean to do this... Is there anyway to clean this mess I just made? You know what? Reject this PR, I'll make a new one if possible.

unknownbrackets replied:
It looks like you rebased from an older commit and maybe merged.  Don't worry, this is easy to fix.

```sh
git checkout local
git remote add upstream https://github.com/hrydgard/ppsspp
git fetch upstream
git rebase -i upstream/master
# Delete the lines with any commits that aren't yours.
git push origin -f local
```

It's important to use the `-f` flag on push and not to do any `git merge` or `git pull` while rebasing.

-[Unknown]

Antx64 replied:
Awesome, that fixed it. Thanks for the help.

unknownbrackets replied:
Cool, but the commits that added and then removed the large binary files are still there, so the large files (if this was merged) would always bloat the size of the .git directory.

You can use `squash` in the rebase instruction list instead of `pick` to combine commits, or else `edit` to change them.

-[Unknown]

Antx64 replied:
Ok, so I found some time now to implement a prebuild command to copy files over. What I've been trying to do for a few hours is to fix the rebase. Whenever I run the rebase -i command, pick the commit I want and squash the others and save, it complains that it doesn't find the file ".git/rebase-merge/git-rebase-todo". If anyone has encountered that error and knows how to fix it, please let me know.

vnctdj replied:
Maybe [this](https://stackoverflow.com/questions/17140570) could help?

Antx64 replied:
Thanks @vnctdj that helped. I could rebase the commits, though, I destroyed my working copy. I'll fix it as soon as I get back home, commit latest changes. And hopefully fully comply with the requested changes.

vnctdj replied:
In case you don't know, deleted commits are not really "destroyed", they are kept back in the `reflog` for quite a long time :)

Antx64 replied:
Perfect, all seemed to have committed now. please, let me know if this fulfill the request. Thanks for the help

unknownbrackets commented:
Looking here:
https://github.com/hrydgard/ppsspp/pull/12031/files

I still see changes to submodules that probably aren't intended.  No reason this pull would affect assets/lang.

We can see these were introduced by 3fe34f71b35f8e2d1d750e00a369ad1aaed54757, so you can mark that "edit" in the rebase, and then reverse the changes to the submodules.  For example:

```sh
pushd assets/lang
git checkout e1cd0446bd5683f0f8a7fb08777027d42aa20e00
popd
pushd ext/armips
git checkout 7885552b208493a6a0f21663770c446c3ba65576
popd
pushd ext/glslang
git checkout 9983f99e87ab0b6608b236ea59bcf873f90e1435
popd
pushd ffmpeg
git checkout 90701640c7f458461310b54e7d4041230e2d5d5a
popd
pushd pspautotests
git checkout 32d3e0426af6fccd67fe00fb4a345e6d2f3ea1a4
popd
git commit --amend assets/lang ext/armips ext/glslang ffmpeg pspautotests
git rebase --continue
```

That will edit the commit to no longer move those submodules to older, more outdated versions.  Doing this may fix the Travis errors which show that, if we merged this pull as is, Android builds would all be broken.  This is because glslang was pushed back to an older version (again, I'm sure this wasn't intentional.)

-[Unknown]

Antx64 commented:
Great. That definitely fixed it. Thanks everyone for the help.

hrydgard commented on UWP/UWP.vcxproj.filters:
```diff
@@ -110,6 +110,88 @@
       <Filter>Content</Filter>
     </None>
     <None Include="PPSSPP_UWP_TemporaryKey.pfx" />
+<<<<<<< HEAD
```
Diff markers left behind, this won't compile.

Antx64 replied:
Damn, I didn't catch that since for some reason my code compiles fine. I'll check as soon as I get home and fix it. Is there are another observation or does the other changes check out and you're ok with them for now?

Antx64 replied:
Fixed this on latest commit

hrydgard commented on Core/Util/GameManager.cpp:
```diff
@@ -223,7 +224,7 @@ bool GameManager::InstallGame(std::string zipfile, bool deleteAfter) {
 		// Note that we do NOT write files that are not in a directory, to avoid random
 		// README files etc.
-		if (strchr(fn, '/') != 0) {
+		if (strchr(fn, '/') != 0 || strchr(fn, '/') == 0) {
```
This line pretty much becomes  ```if (true) ``` and is just useless. This should make you think about why it's there in the first place. For example, the above comment becomes invalid in the case of a ISO or CSO.

I think the logic overall needs cleaning up.

Antx64 replied:
Ok, I'll work on the implementation I talked about in a previous comment then and recommit. What I'm thinking is, leaving the eboot behavior as was originally designed, and only do something else if its a a iso/cso. if you prefer this course of action, please let me know, Thanks.

Antx64 replied:
I cleaned the logic, while leaving the original intent for pbp files

Antx64 commented:
Pushed a commit to address the issues pointed to me. What do you guys think? Is the PR in a mergeable/closing state now? Does everything look good on your side? Let me know, thanks everyone!

LunaMoo commented:
Would be nice to remove duplicate gamecontrollerdb.txt and compat.ini instead of updating them.
 Also remove unecessary and ugly changes to style like for example https://github.com/hrydgard/ppsspp/pull/12031/files#diff-433502ac2faec5aba6941302959897b8L102
 You still pollute history with things you later revert like https://github.com/hrydgard/ppsspp/pull/12031/commits/80dd62109d582ea6a0ac7541655718e5cc25fba8#diff-76fea491fb95525eebbc7df47308a774L113
 And ARMIPS submodule change is still there.
 
 Also take note that PBP is a container, it doesn't have to be named EBOOT.PBP to run on PPSSPP, it can contain whole disc(common format for legally obtained games), a homebrew with all it's assets packed together or just an executable file, you might want to consider that instead of concentrating on other containers vs eboots and just threat all containers equally ewentually keeping special treatment for PBP named as eboot.
 ELF and PRX are also a thing and can be ran on PPSSPP, through those are just executable files.

Antx64 commented:
Hi, I just rebase everything and went to earlier merge of master to fix the files that were appearing modified and weren't supposed to, meaning ARMIPS. I fixed the code format that got replaced and didn't realized after being pointed out to me, when I broke my working copy and a mix of copy/paste and VS after rebasing. Also, gamecontrollerdb.txt and compat.ini wont update again and I also went ahead and deleted all the files inside of the content folder and adds languages and fonts to UWP. I'm sorry for all this, after all, it's my first time trying to contribute to an open source project and I'm not that familiarized with how things work, however, I've learned a lot here trying to get this approved.

Now, regarding the code changes, I think there might be some misunderstanding. The check for EBOOT.pbp was something that was originally there as seen on [66dc0ea] commit.(https://github.com/Antx64/ppsspp/commit/66dc0ea62fc721eebe6b380aec3b659a8b2f0a8f) I did what I mentioned to @hrydgard [here](https://github.com/hrydgard/ppsspp/pull/12031#discussion_r284020060), which hopefully he's ok with, as I didn't want to mess too much with the purpose it was intended for, and as is, it already unpacks everything if one of the file formats are identified. As it is, this PR already contains a lot of fixes. Support for the other formats mentioned, I would gladly develop and contribute to, but I think that for those a new issue should be opened and worked on separately (feel free to assign it to me), as well as make a more robust file format handling implementation, because that's outside of the scope of this initial PR. But that's just my opinion. Let me know what you think.

PS: Most of the commits made here, were mostly to fix the thousands mistakes made by me on git XD.

dominater01 commented:
this would be nice to have merged

so many fixes that make it so much easier to just compile then before

every release i always have to fix it

dominater01 commented:
so this is not gonna go through is it?

hrydgard commented:
It is in a modified form, I've been meaning to rework it a bit but haven't had the time recently. Soon.

hrydgard commented:
I took the UWP stuff and put in a separate pull request, #12074.

Later I'll do the ISO/CSO support.

LunaMoo commented:
Just to note https://github.com/hrydgard/ppsspp/issues/12057#issuecomment-495229559 as the issue while not that big will also affect this.

Basically commercial games should not be stored in game folder(we can fail detection on patched games and generate fakeID for it), just as homebrew shouldn't be stored outside of it(we never generate fakeID outside of game folder).

unknownbrackets commented:
Is this still relevant now?

-[Unknown]

Antx64 commented:
I don't think so. You can close this or remove it, I think

Antx64 commented:
Thanks for working on this! Greatly appreciate it
